src/net/fuchsia/mod.rs

```diff
@@ -0,0 +1,369 @@
+            (READABLE | ERROR | HUP)) != 0
+        {
+            Async::Ready(())
+        } else {
+            self.need_read();
+            Async::NotReady
+        }
+    }
+
+    /// Tests to see if this resource is ready to be written to.
+    /// If it is not, it arranges for the current task to receive a notification
+    /// when a "writable" signal arrives.
+    pub fn poll_write(&self) -> Async<()> {
+        let receiver = self.signal_receiver.receiver();
+        if (receiver.signals.load(Ordering::SeqCst) &
+            (WRITABLE | ERROR | HUP)) != 0
+        {
+            Async::Ready(())
+        } else {
+            self.need_write();
+            Async::NotReady
+        }
+    }
+
+    // Returns a reference to the underlying IO object.
+    pub fn get_ref(&self) -> &T {
+        &self.inner
+    }
+
+    // Returns a mutable reference to the underlying IO object.
+    pub fn get_mut(&mut self) -> &mut T {
+        &mut self.inner
+    }
+
+    /// Arranges for the current task to receive a notification when a "readable"
+    /// signal arrives.
+    pub fn need_read(&self) {
+        let receiver = self.signal_receiver.receiver();
+        receiver.read_task.register();
+        let old = receiver.signals.fetch_and(!READABLE, Ordering::SeqCst);
+        // We only need to schedule a new packet if one isn't already scheduled.
+        // If READABLE was already false, a packet was already scheduled.
+        if (old & READABLE) != 0 {
+            self.schedule_packet(READABLE);
+        }
+    }
+
+    /// Arranges for the current task to receive a notification when a "writable"
+    /// signal arrives.
+    pub fn need_write(&self) {
+        let receiver = self.signal_receiver.receiver();
+        receiver.write_task.register();
+        let old = receiver.signals.fetch_and(!WRITABLE, Ordering::SeqCst);
+        // We only need to schedule a new packet if one isn't already scheduled.
+        // If WRITABLE was already false, a packet was already scheduled.
+        if (old & WRITABLE) != 0 {
+            self.schedule_packet(WRITABLE);
+        }
+    }
+
+    fn schedule_packet(&self, signals: usize) {
+        unsafe {
+            let (mut raw_handle, mut raw_signals) = mem::uninitialized();
+            syscall::__fdio_wait_begin(
+                self.fdio,
+                signals as u32,
+                &mut raw_handle,
+                &mut raw_signals,
+            );
+
+            let handle = zx::Handle::from_raw(raw_handle);
+            let signals = zx::Signals::from_bits_truncate(raw_signals);
+
+            let res = handle.wait_async_handle(
+                self.signal_receiver.remote().port(),
+                self.signal_receiver.key(),
+                signals,
+                zx::WaitAsyncOpts::Once,
+            );
+
+            // The handle is borrowed, so we cannot drop it.
+            mem::forget(handle);
+            res.expect("Error scheduling EventedFd notification");
```

Comment:
Note: this function cannot return `Result` because `need_read` and `need_write` are infallible.